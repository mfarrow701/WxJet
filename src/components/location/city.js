// @flow
import React, {Component} from 'react';
import PropTypes from 'prop-types';

class City extends Component {

    render() {
        return (
            <h3 className="City">{this.props.value || 'No city specified'}</h3>
        )
    }
}

City.propTypes = {
    value: PropTypes.string,
};

export default City;
