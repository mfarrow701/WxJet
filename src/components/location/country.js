// @flow
import React, {Component} from 'react';
import PropTypes from 'prop-types';

class Country extends Component {

    render() {
        return (
            <p className="Country">{this.props.value || 'No country specified'}</p>
        )
    }
}

Country.propTypes = {
    value: PropTypes.string,
};

export default Country;
