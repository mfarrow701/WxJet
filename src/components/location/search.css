.Search {
  min-height: 100%; }
  .Search button {
    background-color: #333333;
    border: none;
    color: white;
    cursor: pointer;
    margin: 8px 0;
    padding: 12px 20px;
    transition: 0.5s;
    width: 100%; }
    .Search button:focus {
      background-color: darkturquoise;
      outline: none; }
    .Search button:hover {
      background-color: #FFAD17; }
  .Search input[type=search] {
    background-color: #333333;
    border: none;
    color: white;
    margin: 8px 0;
    outline: none;
    padding: 12px 20px;
    transition: 0.5s;
    width: 100%;
    -webkit-appearance: none; }
    .Search input[type=search]:focus {
      background-color: darkturquoise; }
    .Search input[type=search]:hover {
      background-color: #FFAD17; }
    .Search input[type=search]::placeholder {
      color: white; }
    .Search input[type=search]::-webkit-search-cancel-button {
      display: none; }
  .Search .List li {
    padding: 5px 10px; }
    .Search .List li:hover {
      background: #FFAD17;
      cursor: pointer; }
    .Search .List li .List-Content {
      align-items: flex-start;
      flex-direction: column; }
      .Search .List li .List-Content p:nth-child(2) {
        color: #333333; }
