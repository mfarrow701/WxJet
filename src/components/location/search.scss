.Search {
  min-height: 100%;
  button {
    background-color: darkturquoise;
    border:           none;
    color:            white;
    cursor:           pointer;
    margin:           8px 0;
    padding:          12px 20px;
    width:            100%;
    &:focus {
      background-color: teal;
      outline:          none;
    }
    &:hover {
      background-color: teal;
    }
  }
  input[type=search] {
    background-color: darkturquoise;
    border:           none;
    color:            white;
    margin:           8px 0;
    outline:          none;
    padding:          12px 20px;
    width:            100%;
    &:focus {
      background-color: teal;
    }
    &::placeholder {
      color: white;
    }
    &::-webkit-search-cancel-button {
      display: none;
    }
  }
  .Item {
    align-items:    center;
    display:        flex;
    flex-direction: row;
    padding:        5px 10px;
    transition:     background 0.2s ease;
    .Icon {
      background: teal;
      min-height: 50px;
      min-width:  50px;
    }
    .Content {
      display:        flex;
      flex-direction: column;
      padding-left:   10px;
    }
    &:hover {
      background: lightgray;
      cursor:     pointer;
    }
  }
}