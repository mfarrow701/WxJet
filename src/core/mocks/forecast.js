export const forecast = {
    'SiteRep': {
        'Wx': {
            'Param': [{
                'name': 'FDm',
                'units': 'C',
                '$': 'Feels Like Day Maximum Temperature'
            }, {'name': 'FNm', 'units': 'C', '$': 'Feels Like Night Minimum Temperature'}, {
                'name': 'Dm',
                'units': 'C',
                '$': 'Day Maximum Temperature'
            }, {'name': 'Nm', 'units': 'C', '$': 'Night Minimum Temperature'}, {
                'name': 'Gn',
                'units': 'mph',
                '$': 'Wind Gust Noon'
            }, {'name': 'Gm', 'units': 'mph', '$': 'Wind Gust Midnight'}, {
                'name': 'Hn',
                'units': '%',
                '$': 'Screen Relative Humidity Noon'
            }, {'name': 'Hm', 'units': '%', '$': 'Screen Relative Humidity Midnight'}, {
                'name': 'V',
                'units': '',
                '$': 'Visibility'
            }, {'name': 'D', 'units': 'compass', '$': 'Wind Direction'}, {
                'name': 'S',
                'units': 'mph',
                '$': 'Wind Speed'
            }, {'name': 'U', 'units': '', '$': 'Max UV Index'}, {'name': 'W', 'units': '', '$': 'Weather Type'}, {
                'name': 'PPd',
                'units': '%',
                '$': 'Precipitation Probability Day'
            }, {'name': 'PPn', 'units': '%', '$': 'Precipitation Probability Night'}]
        },
        'DV': {
            'dataDate': '2018-04-04T07:00:00Z',
            'type': 'Forecast',
            'Location': {
                'i': '350001',
                'lat': '56.8716',
                'lon': '-4.1969',
                'name': 'A\' BHUIDHEANACH BHEAG',
                'country': 'SCOTLAND',
                'continent': 'EUROPE',
                'elevation': '936.0',
                'Period': [{
                    'type': 'Day',
                    'value': '2018-04-04Z',
                    'Rep': [{
                        'D': 'NNE',
                        'Gn': '38',
                        'Hn': '100',
                        'PPd': '98',
                        'S': '31',
                        'V': 'VP',
                        'Dm': '-4',
                        'FDm': '-11',
                        'W': '24',
                        'U': '2',
                        '$': 'Day'
                    }, {
                        'D': 'NNW',
                        'Gm': '22',
                        'Hm': '78',
                        'PPn': '21',
                        'S': '18',
                        'V': 'EX',
                        'Nm': '-7',
                        'FNm': '-16',
                        'W': '2',
                        '$': 'Night'
                    }]
                }, {
                    'type': 'Day',
                    'value': '2018-04-05Z',
                    'Rep': [{
                        'D': 'W',
                        'Gn': '22',
                        'Hn': '92',
                        'PPd': '56',
                        'S': '16',
                        'V': 'MO',
                        'Dm': '-2',
                        'FDm': '-7',
                        'W': '24',
                        'U': '4',
                        '$': 'Day'
                    }, {
                        'D': 'SE',
                        'Gm': '25',
                        'Hm': '85',
                        'PPn': '63',
                        'S': '20',
                        'V': 'VG',
                        'Nm': '-4',
                        'FNm': '-12',
                        'W': '24',
                        '$': 'Night'
                    }]
                }, {
                    'type': 'Day',
                    'value': '2018-04-06Z',
                    'Rep': [{
                        'D': 'SE',
                        'Gn': '47',
                        'Hn': '92',
                        'PPd': '85',
                        'S': '31',
                        'V': 'MO',
                        'Dm': '1',
                        'FDm': '-8',
                        'W': '27',
                        'U': '2',
                        '$': 'Day'
                    }, {
                        'D': 'SSE',
                        'Gm': '40',
                        'Hm': '97',
                        'PPn': '62',
                        'S': '29',
                        'V': 'PO',
                        'Nm': '0',
                        'FNm': '-8',
                        'W': '18',
                        '$': 'Night'
                    }]
                }, {
                    'type': 'Day',
                    'value': '2018-04-07Z',
                    'Rep': [{
                        'D': 'S',
                        'Gn': '25',
                        'Hn': '92',
                        'PPd': '23',
                        'S': '16',
                        'V': 'GO',
                        'Dm': '3',
                        'FDm': '-2',
                        'W': '7',
                        'U': '2',
                        '$': 'Day'
                    }, {
                        'D': 'WSW',
                        'Gm': '18',
                        'Hm': '95',
                        'PPn': '19',
                        'S': '13',
                        'V': 'GO',
                        'Nm': '-1',
                        'FNm': '-6',
                        'W': '7',
                        '$': 'Night'
                    }]
                }, {
                    'type': 'Day',
                    'value': '2018-04-08Z',
                    'Rep': [{
                        'D': 'E',
                        'Gn': '20',
                        'Hn': '86',
                        'PPd': '38',
                        'S': '11',
                        'V': 'GO',
                        'Dm': '3',
                        'FDm': '-2',
                        'W': '7',
                        'U': '2',
                        '$': 'Day'
                    }, {
                        'D': 'ESE',
                        'Gm': '16',
                        'Hm': '93',
                        'PPn': '38',
                        'S': '11',
                        'V': 'GO',
                        'Nm': '-2',
                        'FNm': '-7',
                        'W': '7',
                        '$': 'Night'
                    }]
                }]
            }
        }
    }
};
export const successful = {};
export const failure = 404;