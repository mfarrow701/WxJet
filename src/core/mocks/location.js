export const locations = {
    'locations': [
        {
            'elevation': '50.0',
            'id': '14',
            'latitude': '54.9375',
            'longitude': '-2.8092',
            'name': 'Carlisle Airport',
            'region': 'nw',
            'unitaryAuthArea': 'Cumbria'
        },
        {
            'elevation': '22.0',
            'id': '26',
            'latitude': '53.3336',
            'longitude': '-2.85',
            'name': 'Liverpool John Lennon Airport',
            'region': 'nw',
            'unitaryAuthArea': 'Merseyside'
        },
        {
            'elevation': '25.0',
            'id': '33',
            'latitude': '60.4322',
            'longitude': '-1.2992',
            'name': 'Scatsta',
            'region': 'os',
            'unitaryAuthArea': 'Shetland Islands'
        }
    ]
};
export const selectedLocation = {
    elevation: "119.0",
    id: "352816",
    latitude: "50.4407",
    longitude: "-5.0028",
    name: "Newquay Cornwall Airport",
    region: "sw",
    unitaryAuthArea: "Cornwall"
};
export const successful = {};
export const failure = 404;