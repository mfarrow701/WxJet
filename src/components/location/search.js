// @flow
import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {locationsAPIRequest, locationSelected} from '../../actions/location.actions';
import {filterLocations} from '../../services/locations.service';
import {generateRandomHexCode} from '../../services/utils.service';
import './search.css';
import Loading from '../loading/loading';

class Search extends Component {
    constructor(props) {
        super(props);
        this.state = {
            filteredLocations: null,
        }
    }

    componentWillMount() {
        this.props.requestLocations();
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.locationsPayload !== null) {
            this.setState({
                filteredLocations: filterLocations(nextProps.locationsPayload.locations, '', 15)
            });
        }
    }

    render() {
        let body;
        if (!this.props.locationsError) {
            if (this.state.filteredLocations === null) {
                body = <Loading />
            } else if (this.state.filteredLocations.length === 0) {
                body = (
                    <Fragment>
                        <input onChange={this.handleSearchChange} placeholder="Search for a location..." type="search"/>
                        <h5>No matching locations</h5>
                    </Fragment>
                )
            } else {
                body = (
                    <Fragment>
                        <input onChange={this.handleSearchChange} placeholder="Search for a location..." type="search"/>
                        {this.state.filteredLocations.map(locationData => {
                            return (
                                <div className="Item" key={locationData.id}
                                     onClick={() => this.props.selectLocation(locationData)}>
                                    <div className="Icon" style={{background: generateRandomHexCode()}}/>
                                    <div className="Content">
                                        <h5>{locationData.name}</h5>
                                        <p>{locationData.unitaryAuthArea}</p>
                                    </div>
                                </div>
                            )
                        })}
                    </Fragment>
                );
            }
        } else {
            body = (
                <Fragment>
                    <h5>Error with the locations API</h5>
                    <button onClick={() => this.props.requestLocations()}>Request locations</button>
                </Fragment>
            )
        }
        return (
            <div className="Search">
                {body}
            </div>
        )
    }

    handleSearchChange = (event) => {
        this.setState({
            filteredLocations: filterLocations(this.props.locationsPayload.locations, event.target.value, 15)
        });
    };
}

const mapStateToProps = state => {
    return {
        locationsPayload: state.locationsReducer.payload,
        locationsError: state.locationsReducer.error
    };
};

const mapDispatchToProps = dispatch => {
    return {
        requestLocations: () => dispatch(locationsAPIRequest()),
        selectLocation: location => {
            let storedLocation = localStorage.getItem('storedLocation');
            if (storedLocation && JSON.parse(storedLocation)) {
                // Clear the previously selected location from local storage
                localStorage.removeItem('storedLocation');
            }
            localStorage.setItem('storedLocation', JSON.stringify(location));
            dispatch(locationSelected(location));
        }
    };
};

Search.propTypes = {
    value: PropTypes.string,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
