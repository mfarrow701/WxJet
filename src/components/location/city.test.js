import React from 'react';
import {shallow} from 'enzyme';
import City from './city';
import {selectedLocation} from '../../core/mocks/location';

describe('City component', () => {
    let component;

    const renderWrapper = () => {
        component = shallow(<City/>);
    };

    beforeEach(() => {
        renderWrapper();
    });

    it('should render the city component after load', () => {
        expect(component).toMatchSnapshot();
        expect(component.find('.City').text()).toEqual('No city specified');
    });

    it('should render the city component with a custom message', () => {
        component.setProps({value: selectedLocation.name});
        expect(component.find('.City').text()).toEqual('Newquay Cornwall Airport');
    });
});
