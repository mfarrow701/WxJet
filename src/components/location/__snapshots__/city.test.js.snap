// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City component should render the city component after load 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <City />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "No city specified",
      "className": "City",
    },
    "ref": null,
    "rendered": "No city specified",
    "type": "h3",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No city specified",
        "className": "City",
      },
      "ref": null,
      "rendered": "No city specified",
      "type": "h3",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
